File: legacy-python/acer_profile/profiles.py
"""Definice profilů a načítání konfigurace (TOML).

Standalone režim: acer-profile vlastní VŠECHNY páky (RAPL + EPP + governor +
GPU), protože power-profiles-daemon je na Swift SFG14-73 rozbitý (EBUSY při
přepínání z performance). powerprofilesctl už NENÍ rozhraní.

Profily:
  eco          -> úsporný (20 W)
  normal       -> běžný   (35 W)
  performance  -> výkon   (45/64 W)
"""

from __future__ import annotations

import logging
import os
import tomllib
from dataclasses import dataclass

log = logging.getLogger(__name__)

CONFIG_PATH = "/etc/acer-profile/profiles.toml"

# akceptované názvy profilů (včetně powerprofilesctl aliasů pro pohodlí)
PROFILE_ALIASES = {
    "power-saver": "eco",
    "balanced": "normal",
    "performance": "performance",
    "eco": "eco",
    "normal": "normal",
    "perf": "performance",
}


@dataclass
class Profile:
    pl1_uw: int | None = None      # long_term sustained (W * 1_000_000)
    pl2_uw: int | None = None      # short_term turbo
    epp: str | None = None         # intel_pstate energy_performance_preference
    governor: str | None = None    # performance / powersave
    gpu_max_mhz: int | None = None # i915 gt_max_freq_mhz
    gpu_boost_mhz: int | None = None


# Defaulty kalibrované pro Swift SFG14-73 (Lunar Lake):
#   RAPL PL1 max hlášen 45 W; PL2 (short) bez max, BIOS default 64 W.
#   i915 GPU: RPn/RP1=800 MHz, RP0/boost=2350 MHz.
DEFAULTS: dict[str, Profile] = {
    "eco": Profile(
        pl1_uw=20_000_000,   # 20 W
        pl2_uw=25_000_000,   # 25 W
        epp="power",
        governor="powersave",
        gpu_max_mhz=800,
        gpu_boost_mhz=800,
    ),
    "normal": Profile(
        pl1_uw=35_000_000,   # 35 W
        pl2_uw=45_000_000,   # 45 W
        epp="balance_power",
        governor="powersave",
        gpu_max_mhz=1600,
        gpu_boost_mhz=1600,
    ),
    "performance": Profile(
        pl1_uw=45_000_000,   # 45 W (RAPL max sustained)
        pl2_uw=64_000_000,   # 64 W (BIOS turbo short-term)
        epp="performance",
        governor="performance",
        gpu_max_mhz=2350,
        gpu_boost_mhz=2350,
    ),
}


def _profile_from_dict(d: dict) -> Profile:
    return Profile(
        pl1_uw=d.get("pl1_uw"),
        pl2_uw=d.get("pl2_uw"),
        epp=d.get("epp"),
        governor=d.get("governor"),
        gpu_max_mhz=d.get("gpu_max_mhz"),
        gpu_boost_mhz=d.get("gpu_boost_mhz"),
    )


def load_profiles() -> dict[str, Profile]:
    profiles = {k: v for k, v in DEFAULTS.items()}
    if not os.path.exists(CONFIG_PATH):
        return profiles
    try:
        with open(CONFIG_PATH, "rb") as fh:
            cfg = tomllib.load(fh)
    except OSError as exc:
        log.error("nelze číst %s: %s", CONFIG_PATH, exc)
        return profiles
    except tomllib.TOMLDecodeError as exc:
        log.error("neplatný TOML %s: %s", CONFIG_PATH, exc)
        return profiles
    for name, section in cfg.items():
        if name not in profiles:
            log.warning("ignoruji neznámý profil %s v konfigu", name)
            continue
        profiles[name] = _profile_from_dict(section)
    log.info("profily načteny z %s", CONFIG_PATH)
    return profiles


def resolve(name: str, profiles: dict[str, Profile]) -> Profile | None:
    key = PROFILE_ALIASES.get(name, name)
    return profiles.get(key)


def canonical(name: str) -> str | None:
    return PROFILE_ALIASES.get(name, name) if name in PROFILE_ALIASES else None

File: src/profiles.rs
//! Definice profilů a načítání konfigurace (TOML).
//!
//! Standalone režim: acer-profile vlastní VŠECHNY páky (RAPL + EPP + governor +
//! GPU), protože power-profiles-daemon je na Swift SFG14-73 rozbitý (EBUSY).
//!
//! Profily: eco (20W), normal (35W), performance (45/64W).

use std::collections::HashMap;
use std::fs;
use std::path::Path;

use serde::Deserialize;

pub const CONFIG_PATH: &str = "/etc/acer-profile/profiles.toml";

/// Akceptované názvy profilů (včetně powerprofilesctl aliasů pro pohodlí).
/// Mapuje vstupní název -> kanonický (eco/normal/performance).
pub fn alias(name: &str) -> Option<&'static str> {
    match name {
        "power-saver" | "eco" => Some("eco"),
        "balanced" | "normal" => Some("normal"),
        "performance" | "perf" => Some("performance"),
        _ => None,
    }
}

/// Všechny aliasy (vstupní název -> kanonický) pro nápovědu CLI.
pub const PROFILE_ALIASES_LIST: &[(&str, &str)] = &[
    ("power-saver", "eco"),
    ("balanced", "normal"),
    ("performance", "performance"),
    ("eco", "eco"),
    ("normal", "normal"),
    ("perf", "performance"),
];

/// PPD-style název pro DBus (Plasma očekává power-saver/balanced/performance).
pub fn to_ppd_name(canonical: &str) -> &'static str {
    match canonical {
        "eco" => "power-saver",
        "normal" => "balanced",
        "performance" => "performance",
        _ => "balanced",
    }
}

#[derive(Debug, Clone, Deserialize, Default)]
pub struct Profile {
    #[serde(default)]
    pub pl1_uw: Option<i64>, // long_term sustained (W * 1_000_000)
    #[serde(default)]
    pub pl2_uw: Option<i64>, // short_term turbo
    #[serde(default)]
    pub epp: Option<String>, // intel_pstate energy_performance_preference
    #[serde(default)]
    pub governor: Option<String>, // performance / powersave
    #[serde(default)]
    pub gpu_max_mhz: Option<i64>, // i915 gt_max_freq_mhz
    #[serde(default)]
    pub gpu_boost_mhz: Option<i64>,
}

// Defaulty kalibrované pro Swift SFG14-73 (Meteor Lake-H / Core Ultra 9 185H):
//   RAPL PL1 max hlášen 45 W; PL2 (short) bez max, BIOS default 64 W.
//   i915 GPU: RPn/RP1=800 MHz, RP0/boost=2350 MHz.
fn defaults() -> HashMap<&'static str, Profile> {
    let mut m = HashMap::new();
    m.insert(
        "eco",
        Profile {
            pl1_uw: Some(20_000_000),
            pl2_uw: Some(25_000_000),
            epp: Some("power".to_string()),
            governor: Some("powersave".to_string()),
            gpu_max_mhz: Some(800),
            gpu_boost_mhz: Some(800),
        },
    );
    m.insert(
        "normal",
        Profile {
            pl1_uw: Some(35_000_000),
            pl2_uw: Some(45_000_000),
            epp: Some("balance_power".to_string()),
            governor: Some("powersave".to_string()),
            gpu_max_mhz: Some(1600),
            gpu_boost_mhz: Some(1600),
        },
    );
    m.insert(
        "performance",
        Profile {
            pl1_uw: Some(45_000_000),
            pl2_uw: Some(64_000_000),
            epp: Some("performance".to_string()),
            governor: Some("performance".to_string()),
            gpu_max_mhz: Some(2350),
            gpu_boost_mhz: Some(2350),
        },
    );
    m
}

/// Načte profily: defaulty + override z CONFIG_PATH (zachová neznámé ignoruje).
pub fn load_profiles() -> HashMap<String, Profile> {
    let mut out: HashMap<String, Profile> = HashMap::new();
    for (k, v) in defaults() {
        out.insert(k.to_string(), v);
    }
    let path = Path::new(CONFIG_PATH);
    if !path.exists() {
        return out;
    }
    match fs::read_to_string(path) {
        Ok(content) => match toml::from_str::<HashMap<String, Profile>>(&content) {
            Ok(cfg) => {
                for (name, section) in cfg {
                    if !out.contains_key(&name) {
                        log::warn!("ignoruji neznámý profil {} v konfigu", name);
                        continue;
                    }
                    out.insert(name, section);
                }
                log::info!("profily načteny z {}", CONFIG_PATH);
            }
            Err(e) => log::error!("neplatný TOML {}: {}", CONFIG_PATH, e),
        },
        Err(e) => log::error!("nelze číst {}: {}", CONFIG_PATH, e),
    }
    out
}

/// Resolve alias -> profil z načtené mapy.
pub fn resolve<'a>(name: &str, profiles: &'a HashMap<String, Profile>) -> Option<&'a Profile> {
    alias(name).and_then(|key| profiles.get(key))
}

/// Kanonický název, nebo None pro neznámý.
pub fn canonical(name: &str) -> Option<&'static str> {
    alias(name)
}
